assistant/assistant.go

```diff
@@ -0,0 +1,184 @@
+package assistant
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/influxdata/telegraf/agent"
+	"github.com/influxdata/telegraf/internal"
+)
+
+/*
+Assistant is a client to facilitate communications between Agent and Cloud.
+*/
+type Assistant struct {
+	config     *AssistantConfig // Configuration for Assitant's connection to server
+	connection *websocket.Conn  // Active websocket connection
+	done       chan struct{}    // Channel used to stop server listener
+	agent      *agent.Agent     // Pointer to agent to issue commands
+}
+
+/*
+AssistantConfig allows us to configure where to connect to and other params
+for the agent.
+*/
+type AssistantConfig struct {
+	Host          string
+	Path          string
+	RetryInterval int
+}
+
+func (astConfig *AssistantConfig) fillDefaults() {
+	if astConfig.Host == "" {
+		astConfig.Host = "localhost:8080"
+	}
+	if astConfig.Path == "" {
+		astConfig.Path = "/echo"
+	}
+	if astConfig.RetryInterval == 0 {
+		astConfig.RetryInterval = 15
+	}
+}
+
+// NewAssistant returns an Assistant for the given Config.
+func NewAssistant(ctx context.Context, config *AssistantConfig, agent *agent.Agent) (*Assistant, error) {
+	config.fillDefaults()
+
+	a := &Assistant{
+		config: config,
+		done:   make(chan struct{}),
+		agent:  agent,
+	}
+
+	return a, nil
+}
+
+// Stop is used to clean up active connection and all channels
+func (assistant *Assistant) Stop() {
+	assistant.done <- struct{}{}
+}
+
+type plugin struct {
```

Comment:
nit: does this need to be exported? since the Request will be made by the server

Reply:
I don't think so since the server probably won't be using this go code.

Reply:
Everything related to our plugins will be handled within Assistant -- As you said, this matters if we are using a Go server for the cloud and want to import the Request struct, but since it is such a simple format I think they'll probably construct the JSON from scratch without needing to import anything from our assistant package. I might be mistaken though?